### PR TITLE
[SPARK-31658][SQL] Fix SQL UI not showing write commands of AQE plan

### PR DESCRIPTION
**Pull request by manuzhang**
### What changes were proposed in this pull request?
Show write commands on SQL UI of an AQE plan

### Why are the changes needed?
Currently the leaf node of an AQE plan is always a `AdaptiveSparkPlan` which is not true when it's a child of a write command. Hence, the node of the write command as well as its metrics are not shown on the SQL UI.

#### Before

image

#### After

image

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add UT.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -985,4 +985,27 @@ class AdaptiveQueryExecSuite
...
+          override def onOtherEvent(event: SparkListenerEvent): Unit = {
+            event match {
+              case SparkListenerSQLAdaptiveExecutionUpdate(_, _, planInfo) =>
+                assert(planInfo.nodeName == "Execute CreateDataSourceTableAsSelectCommand")
+                checkDone = true
+              case _ => // ignore other events
+            }
+          }
+        }
+        spark.sparkContext.addSparkListener(listener)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1004]: let's remove the listener at the end of the test

**Review by cloud-fan — APPROVED** [review]
good catch!

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
cc @maryannxue @JkSelf

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 2]
```diff
@@ -524,11 +524,12 @@ case class AdaptiveSparkPlanExec(
       context.session.sparkContext.listenerBus.post(SparkListenerSQLAdaptiveSQLMetricUpdates(
         executionId.toLong, newMetrics))
     } else {
+      val queryExecution = SQLExecution.getQueryExecution(executionId)
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:527]: Use `context.qe` instead??

**Comment by manuzhang** [comment]
@cloud-fan @maryannxue updated. Please review again.

**Review by maryannxue — APPROVED** [review]
LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 3]
```diff
@@ -524,11 +524,12 @@ case class AdaptiveSparkPlanExec(
       context.session.sparkContext.listenerBus.post(SparkListenerSQLAdaptiveSQLMetricUpdates(
         executionId.toLong, newMetrics))
     } else {
+      val queryExecution = context.qe
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:527]: note: `queryExecution` is longer than `context.qe`, we can just inline it.

**Review by cloud-fan — APPROVED** [review]

**Comment by Ngone51** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 4]
```diff
@@ -1004,10 +1004,13 @@ class AdaptiveQueryExecSuite
...
         spark.sparkContext.addSparkListener(listener)
-        sql("CREATE TABLE t1 AS SELECT 1 col").collect()
-        spark.sparkContext.listenerBus.waitUntilEmpty()
-        assert(checkDone)
-        spark.sparkContext.removeSparkListener(listener)
+        try {
+          sql("CREATE TABLE t1 AS SELECT 1 col").collect()
+          spark.sparkContext.listenerBus.waitUntilEmpty()
+          assert(checkDone)
+        } catch {
```
Comment (maryannxue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1011]: finally??

**Comment by gatorsmile** [comment]
Thanks! Merged to master/3.0

**Pull request closed without merging**